airflow/providers_manager.py

```diff
@@ -0,0 +1,78 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Manages all providers."""
+import importlib
+import logging
+import pkgutil
+import traceback
+
+import yaml
+
+try:
+    import importlib.resources as importlib_resources
+except ImportError:
+    # Try backported to PY<37 `importlib_resources`.
+    import importlib_resources
+
+
+log = logging.getLogger(__name__)
+
+
+class ProvidersManager:
+    """Manages all provider packages."""
+
+    def __find_all_providers(self, paths: str):
+        def onerror(_):
+            exception_string = traceback.format_exc()
+            log.warning(exception_string)
+
+        for module_info in pkgutil.walk_packages(paths, prefix="airflow.providers.", onerror=onerror):
+            try:
+                imported_module = importlib.import_module(module_info.name)
+            except Exception as e:  # noqa pylint: disable=broad-except
+                log.warning("Error when importing %s:%s", module_info.name, e)
+                continue
+            try:
+                provider_info = yaml.safe_load(
+                    importlib_resources.read_text(imported_module, 'provider.yaml')
+                )
+                # TODO(potiuk): map to a class maybe? Or we might stick to a dictionary
```

Comment:
Can you add JSON schema validation? 
![image](https://user-images.githubusercontent.com/12058428/99383072-ab2a1280-28cd-11eb-8d37-708358c87f7c.png)

Reply:
I think those provider.yaml should be verified at commit time (pre-commit) rather than in runtime

Reply:
If we also want to support [third parties](https://github.com/apache/airflow/pull/12383#discussion_r524970486), we cannot expect these files to have the correct structure. The user can always damage this file. Besides, there are Airflow forks that do no ru pre-commit on  CI and can also create a damaged package.

Reply:
What Kamil said.

Reply:
Agreee. It is now both pre-commit and runtime :). Better safe than sorry.

Reply:
At some point we're just going to end up running our entire test suite in pre-commit :grin: